openforcefield/topology/molecule.py

```diff
@@ -0,0 +1,3423 @@
+        #return hash(tuple(self.to_dict().items()))
+        return hash(self.to_smiles())
+
+    @classmethod
+    def from_dict(cls, molecule_dict):
+        """
+        Create a new Molecule from a dictionary representation
+        
+        Parameters
+        ----------
+        molecule_dict : OrderedDict
+            A dictionary representation of the molecule.
+
+        Returns
+        -------
+        molecule : Molecule
+            A Molecule created from the dictionary representation
+
+        """
+        # This  implementation is a compromise to let this remain as a classmethod
+        mol = cls()
+        mol.initialize_from_dict(molecule_dict)
+        return mol
+        
+    def initialize_from_dict(self, molecule_dict):
+        """
+        Initialize this Molecule from a dictionary representation
+
+        Parameters
+        ----------
+        molecule_dict : OrderedDict
+            A dictionary representation of the molecule.
+        """
+        # TODO: Provide useful exception messages if there are any failures
+
+        self._initialize()
+        #if molecule_dict['name'] != None:
+        self.name = molecule_dict['name']
+        for atom_dict in molecule_dict['atoms']:
+            self._add_atom(**atom_dict)
+
+        # Handle virtual site unit reattachment and molecule tagging
+        for vsite_dict in molecule_dict['virtual_sites']:
+            vsite_dict_units = deepcopy(vsite_dict)
+
+            # Attach units to epsilon term
+            vsite_dict_units['epsilon'] = deserialize_quantity(vsite_dict['epsilon'])
+            vsite_dict_units['sigma'] = deserialize_quantity(vsite_dict['sigma'])
+            vsite_dict_units['charge_increments'] = deserialize_quantity(vsite_dict['charge_increments'])
+            #if not (vsite_dict['epsilon'] is None):
+            #    epsilon_unitless = vsite_dict['epsilon']
+            #    epsilon_unit = getattr(unit, vsite_dict_units['epsilon_unit'])
+            #    vsite_dict_units['epsilon'] = unit.Quantity(epsilon_unitless, epsilon_unit)
+            #del vsite_dict_units['epsilon_unit']
+
+            # Attach units to sigma term (Remember! We don't store rmin_half, just sigma!)
+            #if not (vsite_dict['sigma'] is None):
+            #    sigma_unitless = vsite_dict['sigma']
+            #    sigma_unit = getattr(unit, vsite_dict_units['sigma_unit'])
+            #    vsite_dict_units['sigma'] = unit.Quantity(sigma_unitless, sigma_unit)
+            #del vsite_dict_units['sigma_unit']
+
+            # Attach units to charge_increments
+            #if not (vsite_dict['charge_increments'] is None):
+            #    charge_increments_unitless = vsite_dict['charge_increments']
+            #    charge_increments_unit = getattr(unit, vsite_dict_units['charge_increments_unit'])
+            #    vsite_dict_units['charge_increments'] = unit.Quantity(charge_increments_unitless,
+            #                                      
```

Comment:
Can we mark the APIs for the virtual site creation calls with a warning that states they are experimental and subject to change? I worry that the API choice here is a bit awkward and will lead to a maintenance burden, since we will also need API calls to modify and delete virtual sites, and having so many different individual methods in the public API can make this less clean than I would like.